applications/FluidDynamicsApplication/custom_conditions/navier_stokes_wall_condition.h

```diff
@@ -0,0 +1,754 @@
+        const double A = std::sqrt(data.Normal[0]*data.Normal[0]+data.Normal[1]*data.Normal[1]+data.Normal[2]*data.Normal[2]);
+        data.Normal /= A;
+
+        // Gauss point information
+        GeometryType& rGeom = this->GetGeometry();
+        const GeometryType::IntegrationPointsArrayType& IntegrationPoints = rGeom.IntegrationPoints(GeometryData::GI_GAUSS_2);
+        const unsigned int NumGauss = IntegrationPoints.size();
+        const MatrixType Ncontainer = rGeom.ShapeFunctionsValues(GeometryData::GI_GAUSS_2);
+
+        for(unsigned int igauss = 0; igauss<NumGauss; igauss++)
+        {
+            data.N = row(Ncontainer, igauss);
+            const double J = rGeom.DeterminantOfJacobian(igauss, GeometryData::GI_GAUSS_2);
+            data.wGauss = J * IntegrationPoints[igauss].Weight();
+
+            ComputeGaussPointRHSContribution(rhs_gauss, data);
+            ComputeGaussPointLHSContribution(lhs_gauss, data);
+
+            noalias(rLeftHandSideMatrix) += lhs_gauss;
+            noalias(rRightHandSideVector) += rhs_gauss;
+        }
+
+        KRATOS_CATCH("")
+    }
+
+    /// Calculates the RHS condition contributions
+    /**
+     * Clones the selected element variables, creating a new one
+     * @param rLeftHandSideMatrix: reference to the LHS matrix
+     * @param rCurrentProcessInfo: reference to the ProcessInfo (unused)
+     */
+    virtual void CalculateLeftHandSide(MatrixType& rLeftHandSideMatrix,
+                                       ProcessInfo& rCurrentProcessInfo)
+    {
+        KRATOS_TRY
+
+        constexpr unsigned int MatrixSize = TNumNodes*(TDim+1);
+
+        if (rLeftHandSideMatrix.size1() != MatrixSize)
+            rLeftHandSideMatrix.resize(MatrixSize, MatrixSize, false); //false says not to preserve existing storage!!
+
+        // Struct to pass around the data
+        ConditionDataStruct data;
+
+        // Allocate memory needed
+        bounded_matrix<double,MatrixSize, MatrixSize> lhs_local;
+
+        // Loop on gauss points
+        noalias(rLeftHandSideMatrix) = ZeroMatrix(MatrixSize,MatrixSize);
+
+        // Compute condition normal
+        this->CalculateNormal(data.Normal); //this already contains the area
+        const double A = std::sqrt(data.Normal[0]*data.Normal[0]+data.Normal[1]*data.Normal[1]+data.Normal[2]*data.Normal[2]);
+        data.Normal /= A;
+
+        // Gauss point information
+        GeometryType& rGeom = this->GetGeometry();
+        const GeometryType::IntegrationPointsArrayType& IntegrationPoints = rGeom.IntegrationPoints(GeometryData::GI_GAUSS_2);
+        const unsigned int NumGauss = IntegrationPoints.size();
+        const MatrixType Ncontainer = rGeom.ShapeFunctionsValues(GeometryData::GI_GAUSS_2);
+
+        for(unsigned int igauss = 0; igauss<NumGauss; igauss++)
+        {
+            data.N = row(Ncontainer, igauss);
+            const double J = rGeom.DeterminantOfJacobian(igauss, GeometryData::GI_GAUSS_2);
```

Comment:
i would suggest that you call the function that computes all the determinants at once, since it allows some optimizations. (outside of the loop of course)